tempesta_fw/tls.c

```diff
@@ -379,23 +388,91 @@ tfw_tls_encrypt(struct sock *sk, struct sk_buff *skb, unsigned int limit)
 
 	if (likely(sgt.nents <= AUTO_SEGS_N)) {
 		sgt.sgl = sg;
+		out_sgt.sgl = out_sg;
+		pages = auto_pages;
 	} else {
 		sgt.sgl = kmalloc(sizeof(struct scatterlist) * sgt.nents,
 				  GFP_ATOMIC);
 		if (!sgt.sgl) {
 			T_WARN("cannot alloc TLS encryption scatter list.\n");
-			return -ENOMEM;
+			r = -ENOMEM;
+			goto err_sgt_alloc;
+		}
+		out_sgt.sgl = kmalloc(sizeof(struct scatterlist) * out_sgt.nents,
+				      GFP_ATOMIC);
+		if (!out_sgt.sgl) {
+			T_WARN("cannot alloc TLS encryption scatter list.\n");
+			r = -ENOMEM;
+			goto err_out_sgt_alloc;
+		}
+		pages = kmalloc(sizeof(struct page *) * out_sgt.nents,
+				GFP_ATOMIC);
+		if (!pages) {
+			T_WARN("cannot alloc memory.\n");
+			r = -ENOMEM;
+			goto err_pages_alloc;
 		}
 	}
 	sg_init_table(sgt.sgl, sgt.nents);
+	sg_init_table(out_sgt.sgl, out_sgt.nents);
 
-	for (next = skb, frags = 0; ; ) {
+	for (next = skb, frags = 0, out_frags = 0, elt = 0; ; ) {
+		unsigned int head_data_len = skb_headlen(next);
 		/*
 		 * skb data and tails are already adjusted above,
 		 * so use zero offset and skb->len.
 		 */
 		r = skb_to_sgvec(next, sgt.sgl + frags, 0, next->len);
 
+		if (skb_shinfo(next)->tx_flags & SKBTX_ZEROCOPY_FRAG) {
```

Comment:
Now I think I was wrong for suggesting testing for `SKBTX_ZEROCOPY_FRAG`.

It's still right in a sense that data is shared with something, but both `SKBTX_SHARED_FRAG` and `SKBTX_DEV_ZEROCOPY` bits set means fragments are **user pages**, and they need to be mapped before use. We should revert to `SKBTX_SHARED_FRAG` here, otherwise we may fail badly on any code that uses `MSG_ZEROCOPY`.

There is an example in `skb_copy_ubufs()` of how `kmap_atomic()`/`kunmap_atomic()` are used to get to the data. The function also uses newly-allocated pages efficiently, to the last byte. Maybe we should also consider using similar technique to save memory.